hibernate-core/src/main/java/org/hibernate/jpa/boot/internal/EntityManagerFactoryBuilderImpl.java

```diff
@@ -567,22 +543,446 @@ else if ( keyString.startsWith( COLLECTION_CACHE_PREFIX ) ) {
+				//noinspection unchecked
+				applyJdbcSettings(
+						integrationJdbcUrl,
+						NullnessHelper.coalesceSuppliedValues(
+								() -> ConfigurationHelper.getString( JPA_JDBC_DRIVER, integrationSettingsCopy ),
+								() -> ConfigurationHelper.getString( JPA_JDBC_DRIVER, mergedSettings.configurationValues )
+						),
+						integrationSettingsCopy,
+						mergedSettings
+				);
+
+				// EARLY EXIT!!
+				return;
+			}
+		}
+
+		if ( persistenceUnit.getJtaDataSource() != null ) {
+			applyDataSource(
+					persistenceUnit.getJtaDataSource(),
+					true,
+					integrationSettingsCopy,
+					mergedSettings
+			);
+
+			// EARLY EXIT!!
+			return;
+		}
+
+		if ( persistenceUnit.getNonJtaDataSource() != null ) {
+			applyDataSource(
+					persistenceUnit.getNonJtaDataSource(),
+					false,
+					integrationSettingsCopy,
+					mergedSettings
+			);
+
+			// EARLY EXIT!!
+			return;
+		}
+
+		if ( mergedSettings.configurationValues.containsKey( URL ) ) {
+			final Object url = mergedSettings.configurationValues.get( URL );
+
+			if ( url != null && ( ! ( url instanceof String ) || StringHelper.isNotEmpty( (String) url ) ) ) {
+				applyJdbcSettings(
+						url,
+						ConfigurationHelper.getString( DRIVER, mergedSettings.configurationValues ),
+						integrationSettingsCopy,
+						mergedSettings
+				);
+
+				// EARLY EXIT!!
+				return;
+			}
+		}
+
+		if ( mergedSettings.configurationValues.containsKey( JPA_JDBC_URL ) ) {
+			final Object url = mergedSettings.configurationValues.get( JPA_JDBC_URL );
+
+			if ( url != null && ( ! ( url instanceof String ) || StringHelper.isNotEmpty( (String) url ) ) ) {
+				applyJdbcSettings(
+						url,
+						ConfigurationHelper.getString( JPA_JDBC_DRIVER, mergedSettings.configurationValues ),
+						integrationSettingsCopy,
+						mergedSettings
+				);
+
+				// EARLY EXIT!!
+				return;
+			}
+		}
+
+		// any other conditions to account for?
+	}
+
+	@SuppressWarnings("unchecked")
+	private void applyDataSource(
+			Object dataSourceRef,
+			Boolean useJtaDataSource,
+			HashMap<?, ?> integrationSettingsCopy,
+			MergedSettings mergedSettings) {
+
+		// `IS_JTA_TXN_COORD` is a value set during `#normalizeTransactionCoordinator` to indicate whether
+		// the execution environment "is JTA" as best as it can tell..
+		//
+		// we use this value when JTA was not explicitly specified in regards the DataSource
+		final boolean isJtaTransactionCoordinator = (boolean) mergedSettings.configurationValues.remove( IS_JTA_TXN_COORD );
+		final boolean isJta = useJtaDataSource == null ? isJtaTransactionCoordinator : useJtaDataSource;
+
+		// add to EMF properties (questionable - see HHH-13432)
+		final String emfKey;
+		final String inverseEmfKey;
+		if ( isJta ) {
+			emfKey = JPA_JTA_DATASOURCE;
+			inverseEmfKey = JPA_NON_JTA_DATASOURCE;
+		}
+		else {
+			emfKey = JPA_NON_JTA_DATASOURCE;
+			inverseEmfKey = JPA_JTA_DATASOURCE;
+		}
+		mergedSettings.configurationValues.put( emfKey, dataSourceRef );
```

Comment:
It seems that due to this change, settings from the `persistence.xml` cannot be cleared anymore, only overridden:

i.e.:

```java
properties.put(AvailableSettings.JPA_JTA_DATASOURCE, null);
entityManagerFactory = Persistence.createEntityManagerFactory("athena", properties);
```

Used to work, but now throws:

```
java.lang.NullPointerException
	at java.util.concurrent.ConcurrentHashMap.putVal(ConcurrentHashMap.java:1011)
	at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1006)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.applyDataSource(EntityManagerFactoryBuilderImpl.java:895)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.normalizeDataAccess(EntityManagerFactoryBuilderImpl.java:758)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.normalizeSettings(EntityManagerFactoryBuilderImpl.java:561)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.mergeSettings(EntityManagerFactoryBuilderImpl.java:485)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:205)
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.<init>(EntityManagerFactoryBuilderImpl.java:175)
	at org.hibernate.jpa.boot.spi.Bootstrap.getEntityManagerFactoryBuilder(Bootstrap.java:76)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilder(HibernatePersistenceProvider.java:171)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:119)
	at org.hibernate.jpa.HibernatePersistenceProvider.getEntityManagerFactoryBuilderOrNull(HibernatePersistenceProvider.java:61)
	at org.hibernate.jpa.HibernatePersistenceProvider.createEntityManagerFactory(HibernatePersistenceProvider.java:50)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:79)
	at com.pallasathenagroup.entitymanagers.AbstractEntityManagerTest.initializeEntityManagerFactory(AbstractEntityManagerTest.java:83)
```

Reply:
Ah nvm , it seems that setting the properties to `null` is simply not required anymore, as the settings are already cleared when setting `AvailableSettings.JPA_JDBC_URL`